docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+boolean fields `parseExtensions` and/or `parseData`. Future work may
+eliminate the need for these fields by detecting the use of extension attributes
+and/or data fields in the expression.
+
+### Why CEL?
+
+CEL is the
+[Common Expression Language](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/doc/langdef.md).
+It is designed for expressing security policy and protocols, and has the
+following
+[guiding philosophy](https://github.com/google/cel-spec/blob/9cdb3682ba04109d2e03d9b048986bae113bf36f/README.md):
+
+*   **Small and fast.** CEL is not Turing complete, so it is easy to understand
+    and implement.
+*   **Extensible.** CEL is designed to be embedded and can be extended with new
+    functionality.
+*   **Developer-friendly.** The language spec was based on experience and
+    usability testing from implementing Firebase Rules.
+
+Additional properties of CEL that make it attractive for Trigger Filtering
+include:
+
+*   **Familiar expression syntax.** Filter expressions written in CEL look and
+    behave like conditional expressions written in the most popular languages.
+*   **Expressions can be compiled and cached for efficiency.** We expect users
+    will create many filters that change rarely in comparison to frequency of
+    evaluations. This usage pattern works in favor of an expression language
+    that can be cached in compiled form.
+*   **Expressions are serializable and composable.** These properties are useful
+    in implementing upstream filter propagation. For example, a Broker could
+    advertise a single CEL expression composed of the union of its Triggers'
+    filters, allowing upstream sources to filter events before publishing to the
+    Broker.
+
+CEL is
+[memory-safe, side-effect-free, and terminating](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview),
+making it safe and secure to embed in a multi-tenant dispatcher process.
+
+## Examples
+
+### Single attribute exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+```
+
+Specified with a CEL expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.type == "com.github.pull.create"
+```
+
+### Multiple attributes exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+      source: https://github.com/knative/eventing/pulls/123
+```
+
+Specified with a CEL expression:
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" &&
+        ce.source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Prefix match on source
+
+Cannot be specified with the SourceAndType field.
+
+CEL doesn't support prefix match directly, but prefix match can be accomplished
```

Comment:
> CEL does support `<string>.startsWith(<prefix>)`

Is `startsWith` a standard definition in the language spec, or an extension added by the Go implementation? I don't see it mentioned in https://github.com/google/cel-spec/blob/master/doc/langdef.md#list-of-standard-definitions. Could we rely on it being present in all runtimes?

I'd worry a bit about `match` (glob) being confused with `matches` (regexp), since the syntaxes are similar but incompatible:

```go
ce.type.match("github*") // good
ce.type.match("github.*") // bad
ce.type.matches("github*") // bad
ce.type.matches("github.*") // good
```

Reply:
Yes, `startsWith` is part of the standard environment. We have yet to update the docs for the standard set of functions, but this should happen once https://github.com/google/cel-go/issues/9 is completed within the CEL Go 0.4.0 milestone.

Reply:
Great! I updated this example to mention `startsWith`, `match`, and `matches`.